frame/crowdloan-rewards/src/lib.rs

```diff
@@ -0,0 +1,325 @@
+		NotInitialized,
+		AlreadyInitialized,
+		InvalidProof,
+		InvalidClaim,
+		NotAContributor,
+		NothingToClaim,
+	}
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		#[allow(missing_docs)]
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+		type Currency: Mutate<Self::AccountId, Balance = Self::Balance>;
+		type Balance: Default
+			+ Parameter
+			+ Codec
+			+ Copy
+			+ Ord
+			+ CheckedAdd
+			+ CheckedSub
+			+ CheckedMul
+			+ AtLeast32BitUnsigned
+			+ MaybeSerializeDeserialize
+			+ Zero;
+		type AdminOrigin: EnsureOrigin<Self::Origin>;
+		type Convert: Convert<Self::BlockNumber, Self::Balance>;
+		type RelayChainAccountId: Parameter + MaybeSerializeDeserialize + Into<AccountId32> + Ord;
+		#[pallet::constant]
+		type InitialPayment: Get<Perbill>;
+		#[pallet::constant]
+		type VestingPartition: Get<Self::BlockNumber>;
+		#[pallet::constant]
+		type Prefix: Get<&'static [u8]>;
+	}
+
+	#[pallet::storage]
+	pub type Rewards<T: Config> =
+		StorageMap<_, Blake2_128Concat, RemoteAccountOf<T>, RewardOf<T>, OptionQuery>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn total_rewards)]
+	pub type TotalRewards<T: Config> = StorageValue<_, T::Balance, ValueQuery>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn claimed_rewards)]
+	pub type ClaimedRewards<T: Config> = StorageValue<_, T::Balance, ValueQuery>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn total_contributors)]
+	pub type TotalContributors<T: Config> = StorageValue<_, u32, ValueQuery>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn vesting_block_start)]
+	pub type VestingBlockStart<T: Config> = StorageValue<_, T::BlockNumber, ValueQuery>;
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		#[pallet::weight(10_000)]
+		pub fn initialize(origin: OriginFor<T>) -> DispatchResult {
+			T::AdminOrigin::ensure_origin(origin)?;
+			ensure!(!VestingBlockStart::<T>::exists(), Error::<T>::AlreadyInitialized);
+			let current_block = frame_system::Pallet::<T>::block_number();
+			VestingBlockStart::<T>::set(current_block);
+			Ok(())
+		}
+
+		#[pallet::weight(10_000)]
+		pub fn claim_from_ethereum(
+			origin: OriginFor<T>,
+			reward_account: T::AccountId,
+			ethereum_address: EthereumAddress,
+			proof: EcdsaSignature,
+		) -> DispatchResult {
+			ensure_signed(origin)?;
+			let data = reward_account.using_encoded(to_ascii_hex);
+			let signer = ethereum_recover(T::Prefix::get(), &proof, &data)
+				.ok_or(Error::<T>::InvalidProof)?;
+			// NOTE(hussein-aitlahcen): ask for the address even if we can extract it
+			// from the signature as we want to differentiate invalid proof from non-existant
+			// contributors
+			ensure!(signer == ethereum_address, Error::<T>::InvalidProof);
+			let remote_account: RemoteAccountOf<T> = RemoteAccount::Ethereum(ethereum_address);
+			let claimed = Self::do_claim(remote_account.clone(), &reward_account)?;
```

Comment:
could you please write down securit scheme?
i read as this 
1. store contirbutors public keys of any kind in storage 
2. during claim, vefify that a. key in strage b. key sin message c. message contains reward account
3. use that reward account to mint into

as for now as i read code, either scheme is differnt or misses some checks from 2 and has non needed checks like substrate native signatures

Reply:
I am not sure I follow what is missing:
1: we setup the rewards in the genesis block, (PK, Amount, Vesting) with PK either ETH/Relay (kusama)
2: you have to sign a message which consist of an arbitrary prefix and the reward account pk (picasso-account_id in tests)
3: you claim by sending the signed message + reward account
4: if that reward account exists, we mint into the reward account

Reply:
ok. i see. it does proper flow. thanks.